File: tools/.mk61-app/mk61_module_pack.cpp
#include "loadable_module_format.hpp"
#include "storage_geometry.hpp"
#include "zx0.hpp"
#include "arm_thumb_bcj.hpp"

#include "third_party/zx0/zx0.h"

#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <limits>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using loadable_module::Compression;
using loadable_module::Header;
using loadable_module::Kind;

static constexpr int ZX0_MAX_OFFSET = 32640;

struct Options {
  Kind kind = Kind::FOCAL;
  bool kind_set = false;
  std::filesystem::path resident;
  std::filesystem::path image;
  std::filesystem::path relocations;
  std::filesystem::path output;
  u32 memory_size = 0;
  u32 entry_offset = 0;
  u32 load_address = 0;
  u16 handled_type_magic = 0;
  bool memory_size_set = false;
  bool entry_offset_set = false;
  bool load_address_set = false;
  bool require_zx0 = false;
  bool portable = false;
};

[[noreturn]] void usage(const char* message = nullptr) {
  if(message != nullptr) std::fprintf(stderr, "error: %s\n\n", message);
  std::fprintf(stderr,
      "usage: mk61_module_pack --kind KIND (--resident FILE | --portable) --image FILE\n"
      "       --memory-size N --entry-offset N --load-address N\n"
      "       --output FILE [--handled-magic XX] [--require-zx0]\n"
      "  --kind KIND          app, focal, tinybasic, wbmp-viewer,\n"
      "                       markdown-viewer, chip8, or setup\n"
      "  --resident FILE      exact resident firmware .bin\n"
      "  --portable           standalone; best of ZX0 / BCJ+ZX0\n"
      "  --relocations FILE   add compact relocation table (ABI 4; otherwise ABI 3)\n"
      "  --image FILE         linked SRAM image without its .bss tail\n"
      "  --memory-size N      image plus zero-filled .bss\n"
      "  --entry-offset N     module entry offset from the load address\n"
      "  --load-address N     linked SRAM base (portable: 0x20000000)\n"
      "  --handled-magic XX   two-byte C5 type magic handled by FILE_OPEN\n"
      "  --require-zx0        reject an uncompressed result\n"
      "  --output FILE        resulting .APP container\n");
  std::exit(message == nullptr ? 0 : 2);
}

u32 parse_u32(const std::string& text, const char* name) {
  std::size_t end = 0;
  unsigned long long value = 0;
  try {
    value = std::stoull(text, &end, 0);
  } catch(const std::exception&) {
    throw std::runtime_error(std::string("invalid ") + name + ": " + text);
  }
  if(end != text.size() || value > std::numeric_limits<u32>::max()) {
    throw std::runtime_error(std::string("invalid ") + name + ": " + text);
  }
  return (u32) value;
}

Kind parse_kind(const std::string& text) {
  if(text == "setup") return Kind::SETUP;
  if(text == "app") return Kind::APPLICATION;
  if(text == "focal") return Kind::FOCAL;
  if(text == "tinybasic") return Kind::TINYBASIC;
  if(text == "wbmp-viewer") return Kind::WBMP_VIEWER;
  if(text == "markdown-viewer") return Kind::MARKDOWN_VIEWER;
  if(text == "chip8") return Kind::CHIP8;
  throw std::runtime_error("unknown module kind: " + text);
}

u16 parse_type_magic(const std::string& text) {
  const auto ascii_alnum = [](unsigned char value) {
    return (value >= '0' && value <= '9') ||
           (value >= 'A' && value <= 'Z') ||
           (value >= 'a' && value <= 'z');
  };
  if(text.size() != 2 || !ascii_alnum((unsigned char) text[0]) ||
     !ascii_alnum((unsigned char) text[1])) {
    throw std::runtime_error(
        "handled magic must contain exactly two ASCII letters or digits");
  }
  return (u16) (u8) text[0] | ((u16) (u8) text[1] << 8);
}

Options parse_options(int argc, char** argv) {
  Options options;
  for(int index = 1; index < argc; index++) {
    const std::string option = argv[index];
    if(option == "--help" || option == "-h") usage();
    if(option == "--require-zx0") {
      options.require_zx0 = true;
      continue;
    }
    if(option == "--portable") {
      options.portable = true;
      options.require_zx0 = true;
      continue;
    }
    if(index + 1 >= argc) usage(("missing value for " + option).c_str());
    const std::string value = argv[++index];
    if(option == "--kind") {
      options.kind = parse_kind(value);
      options.kind_set = true;
    } else if(option == "--resident") {
      options.resident = value;
    } else if(option == "--relocations") {
      options.relocations = value;
    } else if(option == "--image") {
      options.image = value;
    } else if(option == "--memory-size") {
      options.memory_size = parse_u32(value, "memory size");
      options.memory_size_set = true;
    } else if(option == "--entry-offset") {
      options.entry_offset = parse_u32(value, "entry offset");
      options.entry_offset_set = true;
    } else if(option == "--load-address") {
      options.load_address = parse_u32(value, "load address");
      options.load_address_set = true;
    } else if(option == "--handled-magic") {
      options.handled_type_magic = parse_type_magic(value);
    } else if(option == "--output") {
      options.output = value;
    } else {
      usage(("unknown option: " + option).c_str());
    }
  }
  if(!options.portable && !options.relocations.empty()) usage("relocations require --portable");
  if(options.portable) {
    if(!options.resident.empty() ||
       (options.load_address_set &&
        options.load_address != MK61_PORTABLE_APP_ADDRESS)) {
      usage("portable APP requires no resident and fixed SRAM");
    }
    options.load_address = MK61_PORTABLE_APP_ADDRESS;
    options.load_address_set = true;
  }
  if(!options.kind_set || (!options.portable && options.resident.empty()) ||
     options.image.empty() ||
     options.output.empty() || !options.memory_size_set ||
     !options.entry_offset_set || !options.load_address_set) {
    usage("all required options must be specified");
  }
  return options;
}

std::vector<u8> read_file(const std::filesystem::path& path) {
  std::ifstream input(path, std::ios::binary | std::ios::ate);
  if(!input) throw std::runtime_error("cannot open " + path.string());
  const std::streamoff end = input.tellg();
  if(end < 0 || (unsigned long long) end > std::numeric_limits<u32>::max()) {
    throw std::runtime_error("unsupported file size: " + path.string());
  }
  std::vector<u8> bytes((usize) end);
  input.seekg(0);
  if(!bytes.empty() &&
     !input.read((char*) bytes.data(), (std::streamsize) bytes.size())) {
    throw std::runtime_error("cannot read " + path.string());
  }
  return bytes;
}

void write_file(const std::filesystem::path& path,
                const std::vector<u8>& bytes) {
  if(!path.parent_path().empty()) {
    std::filesystem::create_directories(path.parent_path());
  }
  std::ofstream output(path, std::ios::binary | std::ios::trunc);
  if(!output || (!bytes.empty() &&
      !output.write((const char*) bytes.data(),
                    (std::streamsize) bytes.size()))) {
    throw std::runtime_error("cannot write " + path.string());
  }
}

std::vector<u8> zx0_encode_optimal(const std::vector<u8>& input) {
  if(input.empty()) return {};
  if(input.size() > (usize) std::numeric_limits<int>::max()) {
    throw std::runtime_error("ZX0 input is too large");
  }
  std::vector<u8> mutable_input = input;
  const int input_size = (int) mutable_input.size();
  BLOCK* optimal = optimize(mutable_input.data(), input_size, 0,
                            ZX0_MAX_OFFSET);
  if(optimal == nullptr) throw std::runtime_error("ZX0 optimizer failed");
  int output_size = 0;
  int delta = 0;
  unsigned char* output = compress(optimal, mutable_input.data(), input_size,
                                   0, 0, 1, &output_size, &delta);
  if(output == nullptr || output_size <= 0) {
    std::free(output);
    throw std::runtime_error("ZX0 compressor failed");
  }
  std::vector<u8> result(output, output + output_size);
  std::free(output);
  return result;
}

struct VectorInput {
  const std::vector<u8>& bytes;
  usize position = 0;
};

bool next_vector_byte(void* context, u8& value) {
  VectorInput& input = *(VectorInput*) context;
  if(input.position >= input.bytes.size()) return false;
  value = input.bytes[input.position++];
  return true;
}

void verify_zx0(const std::vector<u8>& packed,
                const std::vector<u8>& expected) {
  VectorInput input = {packed, 0};
  const zx0::Input source = {&input, next_vector_byte};
  std::vector<u8> decoded(expected.size());
  u32 written = 0;
  if(!zx0::decode(source, (u32) packed.size(), decoded.data(),
                  (u32) decoded.size(), written) ||
     written != decoded.size() || input.position != packed.size() ||
     decoded != expected) {
    throw std::runtime_error("internal ZX0 verification failed");
  }
}

u32 slot_size(Kind kind) {
  return loadable_module::valid_kind(kind)
      ? loadable_module::MAX_CONTAINER_SIZE : 0;
}

std::vector<u8> pack(const Options& options, const std::vector<u8>& resident,
                     const std::vector<u8>& image) {
  if(!options.portable && (resident.empty() ||
     resident.size() > loadable_module::MAX_RESIDENT_SIZE)) {
    throw std::runtime_error(
        "resident image size is outside the supported range");
  }
  if(image.empty() || image.size() > options.memory_size ||
     options.memory_size > loadable_module::OVERLAY_SIZE) {
    throw std::runtime_error("module image does not fit the SRAM overlay");
  }
  if(options.entry_offset >= image.size() ||
     (options.entry_offset & 1U) != 0) {
    throw std::runtime_error(
        "entry offset must be aligned and inside the image");
  }
  if(options.load_address < loadable_module::SRAM_FIRST_ADDRESS ||
     options.load_address > loadable_module::SRAM_LAST_ADDRESS -
                                loadable_module::OVERLAY_SIZE ||
     (options.load_address & 7U) != 0) {
    throw std::runtime_error(
        "load address is outside aligned STM32F401/F411 SRAM");
  }

  std::vector<u8> stored = zx0_encode_optimal(image);
  verify_zx0(stored, image);
  u32 flags = options.portable ? MK61_PORTABLE_APP_FLAG : 0;
  if(options.portable) {
    std::vector<u8> filtered = image;
    arm_thumb_bcj::transform(filtered.data(), (u32) filtered.size(), true);
    std::vector<u8> candidate = zx0_encode_optimal(filtered);
    verify_zx0(candidate, filtered);
    arm_thumb_bcj::transform(filtered.data(), (u32) filtered.size(), false);
    if(filtered != image) throw std::runtime_error("BCJ round-trip failed");
    // Same header size; ties keep the plain stream and skip the inverse pass.
    std::printf("ZX0 candidates: plain=%zu BCJ=%zu; selected=%s\n",
                stored.size(), candidate.size(),
                candidate.size() < stored.size() ? "BCJ" : "plain");
    if(candidate.size() < stored.size()) {
      stored.swap(candidate);
      flags |= MK61_APP_ARM_THUMB_BCJ_FLAG;
    }
  }
  Compression compression = Compression::ZX0;
  if(!options.require_zx0 && stored.size() >= image.size()) {
    stored = image;
    compression = Compression::NONE;
  }

  const u32 code_stored_size = (u32) stored.size();
  u32 relocation_count = 0;
  if(!options.relocations.empty()) {
    // Empty tables are valid for position-independent images.
    std::ifstream file(options.relocations, std::ios::binary);
    if(!file) throw std::runtime_error("cannot read relocation table");
    const std::vector<u8> table((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
    for(u8 byte : table) if(!(byte & 128U)) ++relocation_count;
    stored.insert(stored.end(), table.begin(), table.end());
    flags |= MK61_APP_RELOCATABLE_FLAG;
  }
  const u32 limit = slot_size(options.kind);
  if(stored.size() + loadable_module::HEADER_SIZE > limit) {
    throw std::runtime_error(
        "module uses " +
        std::to_string(stored.size() + loadable_module::HEADER_SIZE) +
        " bytes; slot limit is " + std::to_string(limit));
  }

  Header header = {};
  header.flags = flags;
  header.kind = options.kind;
  header.compression = compression;
  header.load_address = options.load_address;
  header.stored_size = (u32) stored.size();
  header.image_size = (u32) image.size();
  header.memory_size = options.memory_size;
  header.entry_offset = options.entry_offset;
  header.resident_size = (u32) resident.size();
  header.resident_crc32 =
      loadable_module::crc32(resident.data(), resident.size());
  if(flags & MK61_APP_RELOCATABLE_FLAG) {
    header.code_stored_size = code_stored_size;
    header.relocation_count = relocation_count;
  }
  header.stored_crc32 =
      loadable_module::crc32(stored.data(), stored.size());
  header.image_crc32 =
      loadable_module::crc32(image.data(), image.size());
  header.handled_type_magic = options.handled_type_magic;

  const loadable_module::Reader reader = {&stored,
      [](void* context, u32 offset, u8* out, usize size) -> bool {
        const auto& data = *(const std::vector<u8>*) context;
        if(offset > data.size() || size > data.size() - offset) return false;
        std::memcpy(out, data.data() + offset, size); return true;
      }};
  std::vector<u8> verified(image.size());
  if(!loadable_module::decode_image(header, reader, verified.data(), header.load_address) || verified != image)
    throw std::runtime_error("APP decode/relocation verification failed");

  std::vector<u8> module(loadable_module::HEADER_SIZE + stored.size());
  if(!loadable_module::encode_header(header, limit, module.data())) {
    throw std::runtime_error("cannot encode module header");
  }
  std::copy(stored.begin(), stored.end(),
            module.begin() + loadable_module::HEADER_SIZE);
  return module;
}

} // namespace

int main(int argc, char** argv) {
  try {
    const Options options = parse_options(argc, argv);
    const std::vector<u8> resident = options.portable
        ? std::vector<u8>{} : read_file(options.resident);
    const std::vector<u8> image = read_file(options.image);
    const std::vector<u8> module = pack(options, resident, image);
    write_file(options.output, module);
    const double ratio =
        (double) (module.size() - loadable_module::HEADER_SIZE) /
        (double) image.size();
    std::printf("%s: %zu bytes, payload ratio %.1f%%\n",
                options.output.string().c_str(), module.size(), ratio * 100.0);
    return 0;
  } catch(const std::exception& error) {
    std::fprintf(stderr, "mk61_module_pack: %s\n", error.what());
    return 2;
  }
}

File: tools/.mk61-app/mk61_module_pack.py
#!/usr/bin/env python3
"""Dependency-free ABI 5 APP packer used when no native host C++ exists.

The ZX0 stream writer follows the v2 format by Einar Saukas.  The greedy
parser is intentionally implemented in Python so Arduino IDE on Windows only
needs the Python installation that is already required by the APP build.
ZX0 format copyright (c) 2021 Einar Saukas, BSD-3-Clause; the complete notice
is stored in third_party/zx0/LICENSE.
"""

from __future__ import annotations

import argparse
import struct
import sys
import zlib
from pathlib import Path


HEADER_SIZE = 64
APP_MAX_MEMORY_SIZE = 20 * 1024
LOAD_ADDRESS = 0x20000000
MAX_OFFSET = 32640

PORTABLE_FLAG = 1
BCJ_FLAG = 2
RELOCATABLE_FLAG = 4

KINDS = {
    "focal": 1,
    "tinybasic": 2,
    "wbmp-viewer": 3,
    "app": 4,
    "chip8": 5,
    "markdown-viewer": 6,
    "setup": 7,
}


class PackError(ValueError):
    pass


class BitWriter:
    def __init__(self) -> None:
        self.output = bytearray()
        self.bit_index = 0
        self.bit_mask = 0
        # ZX0 stores the first length bit of a new-offset match in the low
        # offset byte.  At stream start the same mechanism omits the implicit
        # first literal selector.
        self.backtrack = True

    def bit(self, value: bool) -> None:
        if self.backtrack:
            if value:
                if not self.output:
                    raise PackError("invalid ZX0 backtrack")
                self.output[-1] |= 1
            self.backtrack = False
            return
        if self.bit_mask == 0:
            self.bit_mask = 0x80
            self.bit_index = len(self.output)
            self.output.append(0)
        if value:
            self.output[self.bit_index] |= self.bit_mask
        self.bit_mask >>= 1

    def byte(self, value: int) -> None:
        self.output.append(value & 0xFF)

    def gamma(self, value: int, *, inverted: bool = False) -> None:
        if value <= 0:
            raise PackError("invalid ZX0 gamma value")
        bit = 1 << (value.bit_length() - 1)
        while True:
            bit >>= 1
            if bit == 0:
                break
            self.bit(False)
            data = bool(value & bit)
            self.bit(not data if inverted else data)
        self.bit(True)


def match_length(data: bytes, position: int, candidate: int) -> int:
    limit = len(data) - position
    length = 2
    while length < limit and data[position + length] == data[candidate + length]:
        length += 1
    return length


def best_match(data: bytes, position: int, last_offset: int) -> tuple[int, int]:
    if position + 1 >= len(data):
        return 0, 0
    needle = data[position:position + 2]
    window = max(0, position - MAX_OFFSET)
    candidate = data.rfind(needle, window, position)
    best_length = 0
    best_offset = 0
    remaining = len(data) - position
    while candidate >= window:
        offset = position - candidate
        length = match_length(data, position, candidate)
        if (length > best_length or
                (length == best_length and
                 (offset == last_offset or
                  (best_offset != last_offset and offset < best_offset)))):
            best_length = length
            best_offset = offset
            if length == remaining:
                break
        candidate = data.rfind(needle, window, candidate)
    return best_length, best_offset


def greedy_tokens(data: bytes) -> list[tuple[int, int]]:
    if not data:
        raise PackError("ZX0 input is empty")
    tokens: list[tuple[int, int]] = []
    literal_start = 0
    position = 1
    last_offset = 1
    while position + 1 < len(data):
        length, offset = best_match(data, position, last_offset)
        # A two-byte match with a new offset normally costs more once the
        # surrounding literal blocks and offset selector are included.  Keep
        # it only for the compact last-offset form; otherwise leaving both
        # bytes in the current literal run produces a smaller stream.
        if length < 2 or (length == 2 and offset != last_offset):
            position += 1
            continue
        # Do not consume a short match when one literal byte exposes a
        # substantially longer one.  This small lazy step recovers most of
        # the size lost by a purely greedy parser without the quadratic RAM
        # use of the desktop optimal parser.
        if position + 2 < len(data):
            next_length, _ = best_match(data, position + 1, last_offset)
            if next_length > length + 1:
                position += 1
                continue
        tokens.append((position - literal_start, 0))
        tokens.append((length, offset))
        last_offset = offset
        literal_start = position + length
        if literal_start >= len(data):
            break
        position = literal_start + 1
    if literal_start < len(data):
        tokens.append((len(data) - literal_start, 0))
    if not tokens or tokens[0][1] != 0:
        raise PackError("invalid ZX0 token plan")
    return tokens


def zx0_encode(data: bytes) -> bytes:
    writer = BitWriter()
    position = 0
    last_offset = 1
    for length, offset in greedy_tokens(data):
        if length <= 0 or position + length > len(data):
            raise PackError("invalid ZX0 token")
        if offset == 0:
            writer.bit(False)
            writer.gamma(length)
            for value in data[position:position + length]:
                writer.byte(value)
        elif offset == last_offset:
            writer.bit(False)
            writer.gamma(length)
        else:
            writer.bit(True)
            writer.gamma((offset - 1) // 128 + 1, inverted=True)
            writer.byte((127 - ((offset - 1) % 128)) << 1)
            writer.backtrack = True
            writer.gamma(length - 1)
            last_offset = offset
        position += length
    if position != len(data):
        raise PackError("incomplete ZX0 token plan")
    writer.bit(True)
    writer.gamma(256, inverted=True)
    return bytes(writer.output)


class BitReader:
    def __init__(self, data: bytes) -> None:
        self.data = data
        self.position = 0
        self.bit_value = 0
        self.bit_mask = 0

    def byte(self) -> int:
        if self.position >= len(self.data):
            raise PackError("truncated ZX0 stream")
        value = self.data[self.position]
        self.position += 1
        return value

    def bit(self) -> int:
        self.bit_mask >>= 1
        if self.bit_mask == 0:
            self.bit_mask = 0x80
            self.bit_value = self.byte()
        return int(bool(self.bit_value & self.bit_mask))

    def gamma(self, *, inverted: bool, first: int | None = None) -> int:
        result = 1
        current = self.bit() if first is None else first
        while current == 0:
            result = (result << 1) | (self.bit() ^ int(inverted))
            current = self.bit()
        return result


def zx0_decode(data: bytes, expected_size: int) -> bytes:
    reader = BitReader(data)
    output = bytearray()
    last_offset = 1
    state = "literal"
    while True:
        if state == "literal":
            length = reader.gamma(inverted=False)
            for _ in range(length):
                output.append(reader.byte())
            selector = reader.bit()
            if selector:
                state = "new"
            else:
                length = reader.gamma(inverted=False)
                if last_offset > len(output):
                    raise PackError("invalid ZX0 last offset")
                for _ in range(length):
                    output.append(output[-last_offset])
                state = "new" if reader.bit() else "literal"
        else:
            high = reader.gamma(inverted=True)
            if high == 256:
                if len(output) != expected_size or reader.position != len(data):
                    raise PackError("invalid ZX0 end marker")
                return bytes(output)
            if high > 255:
                raise PackError("invalid ZX0 offset")
            low = reader.byte()
            last_offset = high * 128 - (low >> 1)
            length = reader.gamma(inverted=False, first=low & 1) + 1
            if last_offset <= 0 or last_offset > len(output):
                raise PackError("invalid ZX0 offset")
            for _ in range(length):
                output.append(output[-last_offset])
            state = "new" if reader.bit() else "literal"
        if len(output) > expected_size:
            raise PackError("ZX0 output exceeds image")


def bcj_transform(data: bytes, encode: bool) -> bytes:
    output = bytearray(data)
    index = 0
    while index + 3 < len(output):
        if ((output[index + 1] & 0xF8) == 0xF0 and
                (output[index + 3] & 0xF8) == 0xF8):
            address = (((output[index + 1] & 7) << 19) |
                       (output[index] << 11) |
                       ((output[index + 3] & 7) << 8) |
                       output[index + 2]) << 1
            address = address + index + 4 if encode else address - index - 4
            address = (address & 0xFFFFFFFF) >> 1
            output[index + 1] = 0xF0 | ((address >> 19) & 7)
            output[index] = (address >> 11) & 0xFF
            output[index + 3] = 0xF8 | ((address >> 8) & 7)
            output[index + 2] = address & 0xFF
            index += 4
        else:
            index += 2
    return bytes(output)


def validate_relocations(table: bytes, image: bytes, memory_size: int) -> int:
    position = 0
    end = 0
    count = 0
    while position < len(table):
        gap = 0
        shift = 0
        while True:
            if position >= len(table) or shift > 14:
                raise PackError("malformed relocation table")
            value = table[position]
            position += 1
            if shift and value == 0:
                raise PackError("non-canonical relocation table")
            gap |= (value & 0x7F) << shift
            if not value & 0x80:
                break
            shift += 7
        offset = end + gap
        if offset + 4 > len(image):
            raise PackError("relocation is outside the image")
        pointer = struct.unpack_from("<I", image, offset)[0]
        if not LOAD_ADDRESS <= pointer <= LOAD_ADDRESS + memory_size:
            raise PackError("relocation does not refer to the APP image")
        end = offset + 4
        count += 1
    return count


def type_magic(value: str | None) -> int:
    if value is None:
        return 0
    if len(value) != 2 or not value.isascii() or not value.isalnum():
        raise PackError("handled magic must contain two ASCII letters or digits")
    return ord(value[0]) | (ord(value[1]) << 8)


def pack(args: argparse.Namespace) -> bytes:
    image = args.image.read_bytes()
    table = args.relocations.read_bytes()
    memory_size = args.memory_size
    if not image or len(image) > memory_size or memory_size > APP_MAX_MEMORY_SIZE:
        raise PackError("module exceeds the 20 KiB APP image limit")
    if args.entry_offset < 0 or args.entry_offset >= len(image) or args.entry_offset & 1:
        raise PackError("entry offset must be aligned and inside the image")
    if args.load_address != LOAD_ADDRESS:
        raise PackError("current APP ABI has a fixed virtual link base")
    relocation_count = validate_relocations(table, image, memory_size)

    plain = zx0_encode(image)
    if zx0_decode(plain, len(image)) != image:
        raise PackError("internal plain ZX0 verification failed")
    filtered = bcj_transform(image, True)
    if bcj_transform(filtered, False) != image:
        raise PackError("BCJ round-trip failed")
    bcj = zx0_encode(filtered)
    if zx0_decode(bcj, len(filtered)) != filtered:
        raise PackError("internal BCJ ZX0 verification failed")
    use_bcj = len(bcj) < len(plain)
    code = bcj if use_bcj else plain
    print(f"ZX0 candidates: plain={len(plain)} BCJ={len(bcj)}; "
          f"selected={'BCJ' if use_bcj else 'plain'}")

    stored = code + table
    if HEADER_SIZE + len(stored) > HEADER_SIZE + APP_MAX_MEMORY_SIZE:
        raise PackError("module exceeds the 20 KiB APP container limit")
    flags = PORTABLE_FLAG | RELOCATABLE_FLAG | (BCJ_FLAG if use_bcj else 0)
    header = bytearray(HEADER_SIZE)
    header[:8] = b"MK61APP\0"
    struct.pack_into("<HHHBBIIIIIIIIIIH", header, 8,
                     1, HEADER_SIZE, 5, KINDS[args.kind], 1, flags,
                     args.load_address, len(stored), len(image), memory_size,
                     args.entry_offset, len(code), relocation_count,
                     zlib.crc32(stored), zlib.crc32(image),
                     type_magic(args.handled_magic))
    struct.pack_into("<H", header, 58, 0)
    struct.pack_into("<I", header, 60, zlib.crc32(header[:60]))
    return bytes(header) + stored


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--portable", action="store_true", help=argparse.SUPPRESS)
    parser.add_argument("--kind", choices=KINDS, required=True)
    parser.add_argument("--image", type=Path, required=True)
    parser.add_argument("--relocations", type=Path, required=True)
    parser.add_argument("--memory-size", type=lambda x: int(x, 0), required=True)
    parser.add_argument("--entry-offset", type=lambda x: int(x, 0), required=True)
    parser.add_argument("--load-address", type=lambda x: int(x, 0),
                        default=LOAD_ADDRESS)
    parser.add_argument("--handled-magic")
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    try:
        module = pack(args)
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_bytes(module)
        ratio = (len(module) - HEADER_SIZE) * 100.0 / args.image.stat().st_size
        print(f"{args.output}: {len(module)} bytes, payload ratio {ratio:.1f}%")
    except (OSError, PackError, KeyError, struct.error) as error:
        parser.exit(2, f"mk61_module_pack: {error}\n")


if __name__ == "__main__":
    main()
